File: CSharpFundamentals/CSharpOOPbasics/DefiningClassesEx/SpeedRacing/Car.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SpeedRacing
{
    public class Car
    {
        private string model;
        private double fuelAmount;
        private double fuelConsumption;
        private int traveledDistance;

        public Car(string model, double fuelAmount, double fuelConsumption, int traveledDistance)
        {
            this.Model = model;
            this.FuelAmount = fuelAmount;
            this.FuelConsumption = fuelConsumption;
            this.TraveledDistance = traveledDistance;
        }

        public int TraveledDistance
        {
            get { return traveledDistance; }
            set { traveledDistance = value; }
        }

        public double FuelConsumption
        {
            get { return fuelConsumption; }
            set { fuelConsumption = value; }
        }

        public double FuelAmount
        {
            get { return fuelAmount; }
            set { fuelAmount = value; }
        }

        public string Model
        {
            get { return model; }
            set { model = value; }
        }

        public bool CanMove(int amountOfKm)
        {
            return amountOfKm * (this.FuelConsumption) <= this.FuelAmount;
        }
    }
}

File: JavaFundamentals/JavaAdvanced/DefiningClasses_Ex/src/SpeedRacing/Car.java
package SpeedRacing;

public class Car {
    private String model;
    private double fuelAmount;
    private double fuelConsumption;
    private double distanceTraveled = 0;

    public Car(String model, double fuelAmount, double fuelConsumption) {
        this.model = model;
        this.fuelAmount = fuelAmount;
        this.fuelConsumption = fuelConsumption;
    }

    public String getModel() {
        return model;
    }

    public double getFuelAmount() {
        return fuelAmount;
    }

    public double getFuelConsumption() {
        return fuelConsumption;
    }

    public double getDistanceTraveled() {
        return distanceTraveled;
    }

    private boolean canMove(double distance) {
        return this.getFuelAmount() / this.getFuelConsumption() >= distance;
    }

    public void move(double distance) {
        if (this.canMove(distance)) {
            this.fuelAmount -= distance * this.getFuelConsumption();
            this.distanceTraveled += distance;
        } else {
            System.out.println("Insufficient fuel for the drive");
        }
    }

    @Override
    public String toString() {
        return String.format("%s %.2f %.0f",
                this.getModel(),
                this.getFuelAmount(),
                this.getDistanceTraveled());
    }
}
